openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -130,45 +247,166 @@ def __contains__(self, item):
 
         """
+        import openforcefield.utils.toolkits
+
+        self._COSMETIC_ATTRIBS = []  # A list that may be populated to record the cosmetic
+        # attributes read from a SMIRNOFF data source
+
         if smirks is None:
             raise ValueError("'smirks' must be specified")
+
+        # TODO: Make better switch for toolkit registry
+        if openforcefield.utils.toolkits.OPENEYE_AVAILABLE:
+            toolkit = 'openeye'
+        elif openforcefield.utils.toolkits.RDKIT_AVAILABLE:
+            toolkit = 'rdkit'
+        ChemicalEnvironment.validate(
+            smirks, ensure_valence_type=self._VALENCE_TYPE, toolkit=toolkit)
+
         self._smirks = smirks
 
-        # Handle all unknown kwargs as cosmetic so we can write them back out
+
+        # First look for indexed attribs, removing them from kwargs as they're found
+        for unidx_key in self._INDEXED_ATTRIBS:
+            index = 1
+            idx_key = unidx_key+str(index)
+            if idx_key in kwargs:
+                setattr(self, unidx_key, list())
+            while idx_key in kwargs:
+                val = kwargs[idx_key]
+                if unidx_key in self._REQUIRE_UNITS:
+                    if not val.unit.is_compatible(self._REQUIRE_UNITS[unidx_key]):
+
+                        msg = "{} constructor received kwarg {} with value {}, " \
+                              "which is incompatible with expected unit {}".format(self.__class__,
+                                                                                   idx_key,
+                                                                                   val,
+                                                                                   self._REQUIRE_UNITS[unidx_key])
+                        raise SMIRNOFFSpecError(msg)
+                if unidx_key in self._ATTRIBS_TO_TYPE:
+                    type_to_cast = self._ATTRIBS_TO_TYPE[unidx_key]
+                    val = type_to_cast(val)
+                getattr(self, unidx_key).append(val)
+                del kwargs[idx_key]
+                index += 1
+                idx_key = unidx_key+str(index)
+
+
+
         for key, val in kwargs.items():
-            attr_name = '_' + key
-            setattr(self, attr_name, val)
+            if key in self._REQUIRE_UNITS:
+                if not val.unit.is_compatible(self._REQUIRE_UNITS[key]):
+                    msg = "{} constructor received kwarg {} with value {}, " \
+                          "which is incompatible with expected unit {}".format(self.__class__,
+                                                                               key,
+                                                                               val,
+                                                                               self._REQUIRE_UNITS[key])
+                    raise SMIRNOFFSpecError(msg)
+
+                if key in self._ATTRIBS_TO_TYPE:
```

Comment:
Should this be indented into the `if key in self._REQUIRE_UNITS` block? In the statement just above `if key in self._ATTRIBS_TO_TYPE` is its own independent if statement.

Reply:
You might wanter to consider refactoring this check into a _validate_kwarg method to remove the duplicate code?

Reply:
> Should this be indented into the if key in self._REQUIRE_UNITS block?
No, this is unfortunately spaghetti code at the moment, but it's readable. 

I like your validate idea -- Implemented it for units in d062fdc